### PR TITLE
T5530: isis: Adding loop free alternate feature

### PR DESCRIPTION
**Pull request by Cheeze-It**
## Change Summary

We are adding the loop free alternate (LFA) functionality in IS-IS here. This
will allow for said functionality to be enabled as well as a bunch of the 
different options that are available within FRR are exposed to the user.  

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T5530

## Component(s) name

isis

## Proposed changes

We are adding an lfa folder, local and remote, and then options underneath all of them. For the
smoketest we are adding a bunch of tests as well as a lot of these options tend to not be able 
to be configured together as they are mutually exclusive. However everything does seem to 
be able to be configured. Now as for whether or not it actually works properly, I do not know
how to properly test this in a lab. But at least we have the functionality available to us now.

## How to test

Added the config here:
```
vyos@vyos# compare
[policy]
+ prefix-list test-lfa-prefix-list {
+     rule 1 {
+         action "permit"
+         prefix "10.0.0.0/8"
+     }
+ }
[protocols]
+ isis {
+     interface lo {
+     }
+     lfa {
+         remote {
+             prefix-list test-lfa-prefix-list {
+                 level-1
+                 level-2
+             }
+         }
+     }
+     net "49.0001.1921.6800.1002.00"
+ }
```

Showing that the configuration properly worked:
```
vyos@vyos:~$ vtysh -c "show run"
Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
interface eth0
exit
!
interface eth1
exit
!
router isis VyOS
 net 49.0001.1921.6800.1002.00
 fast-reroute remote-lfa prefix-list test-lfa-prefix-list level-1
exit
!
ip prefix-list test-lfa-prefix-list seq 1 permit 10.0.0.0/8
!
rpki
exit
!
end
```

Showing that IS-IS actually turned on loop free alternate (lfa) and the statistics
that surround it:
```
vyos@vyos:~$ show isis fast-reroute summary
Area VyOS:
 IS-IS L1 IPv4 Fast ReRoute summary:

 Protection \ Priority     Critical  High      Medium    Low       Total
 -------------------------------------------------------------------------
 Classic LFA               0         0         0         0         0
 Remote LFA                0         0         0         0         0
 Topology Independent LFA  0         0         0         0         0
 ECMP                      0         0         0         0         0
 Unprotected               0         0         0         0         0
 Protection coverage       0.00%     0.00%     0.00%     0.00%     0.00%

 IS-IS L1 IPv6 Fast ReRoute summary:

 Protection \ Priority     Critical  High      Medium    Low       Total
 -------------------------------------------------------------------------
 Classic LFA               0         0         0         0         0
 Remote LFA                0         0         0         0         0
 Topology Independent LFA  0         0         0         0         0
 ECMP                      0         0         0         0         0
 Unprotected               0         0         0         0         0
 Protection coverage       0.00%     0.00%     0.00%     0.00%     0.00%

 IS-IS L2 IPv4 Fast ReRoute summary:

 Protection \ Priority     Critical  High      Medium    Low       Total
 -------------------------------------------------------------------------
 Classic LFA               0         0         0         0         0
 Remote LFA                0         0         0         0         0
 Topology Independent LFA  0         0         0         0         0
 ECMP                      0         0         0         0         0
 Unprotected               0         0         0         0         0
 Protection coverage       0.00%     0.00%     0.00%     0.00%     0.00%

 IS-IS L2 IPv6 Fast ReRoute summary:

 Protection \ Priority     Critical  High      Medium    Low       Total
 -------------------------------------------------------------------------
 Classic LFA               0         0         0         0         0
 Remote LFA                0         0         0         0         0
 Topology Independent LFA  0         0         0         0         0
 ECMP                      0         0         0         0         0
 Unprotected               0         0         0         0         0
 Protection coverage       0.00%     0.00%     0.00%     0.00%     0.00%
```

Smoketest shows that it passes properly, but the teardown fails. That doesn't seem to
be part of my function though. So I am unsure if that's something that I broke:
```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_protocols_isis.py
test_isis_01_redistribute (__main__.TestProtocolsISIS.test_isis_01_redistribute) ...
Network entity is mandatory!

ok
test_isis_02_vrfs (__main__.TestProtocolsISIS.test_isis_02_vrfs) ... ok
test_isis_04_default_information (__main__.TestProtocolsISIS.test_isis_04_default_information) ... ok
test_isis_05_password (__main__.TestProtocolsISIS.test_isis_05_password) ...
Can use either md5 or plaintext-password for area-password!


Can use either md5 or plaintext-password for domain-password!

ok
test_isis_06_spf_delay_bfd (__main__.TestProtocolsISIS.test_isis_06_spf_delay_bfd) ...
All types of spf-delay must be configured. Missing: init-delay, long-
delay, time-to-learn, short-delay


All types of spf-delay must be configured. Missing: long-delay, time-to-
learn, short-delay


All types of spf-delay must be configured. Missing: short-delay, time-
to-learn


All types of spf-delay must be configured. Missing: time-to-learn

ok
test_isis_07_segment_routing_configuration (__main__.TestProtocolsISIS.test_isis_07_segment_routing_configuration) ... ok
test_isis_08_ldp_sync (__main__.TestProtocolsISIS.test_isis_08_ldp_sync) ... ok
test_isis_09_lfa (__main__.TestProtocolsISIS.test_isis_09_lfa) ... ok
tearDownClass (__main__.TestProtocolsISIS) ... ERROR

======================================================================
ERROR: tearDownClass (__main__.TestProtocolsISIS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/libexec/vyos/tests/smoke/cli/base_vyostest_shim.py", line 57, in tearDownClass
    cls._session.migrate_and_load_config(save_config)
  File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 200, in migrate_and_load_config
    out = self.__run_command(MIGRATE_LOAD_CONFIG + [file_path])
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/vyos/configsession.py", line 139, in __run_command
    raise ConfigSessionError(output)
vyos.configsession.ConfigSessionError: Loading configuration from '/tmp/vyos-smoketest-save'
Traceback (most recent call last):
  File "/usr/libexec/vyos/vyos-load-config.py", line 92, in <module>
    migration.run()
  File "/usr/lib/python3/dist-packages/vyos/migrator.py", line 191, in run
    rev_versions = self.run_migration_scripts(cfg_versions, sys_versions)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/vyos/migrator.py", line 93, in run_migration_scripts
    self.init_logger()
  File "/usr/lib/python3/dist-packages/vyos/migrator.py", line 45, in init_logger
    fh = logging.FileHandler(log_file)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/logging/__init__.py", line 1181, in __init__
    StreamHandler.__init__(self, self._open())
                                 ^^^^^^^^^^^^
  File "/usr/lib/python3.11/logging/__init__.py", line 1213, in _open
    return open_func(self.baseFilename, self.mode,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/opt/vyatta/etc/config/vyos-migrate.log'


----------------------------------------------------------------------
Ran 8 tests in 55.511s

FAILED (errors=1)

```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

* = Once PR approved, will add documentation

### CODE REVIEW

**Comment by c-po** [comment]
Thanks @Cheeze-It for the PR.

Any reason to not have this under `fast-reroute` CLI node as FRR does? So it would be more consistent accross both platforms CLI wise.

**Comment by Cheeze-It** [comment]
> Thanks @Cheeze-It for the PR.
> 
> Any reason to not have this under `fast-reroute` CLI node as FRR does? So it would be more consistent accross both platforms CLI wise.

To be honest with you, it was because of how the CLI was laid out in FRR itself. So both IS-IS and OSPF have a fast-reroute node but inside those nodes there's only one option. IS-IS has "lfa" and OSPF has "ti-lfa". While yes they are accurately portraying that LFA and TI-LFA are indeed fast reroute mechanisms, putting them in a container/node like fast-reroute to me says that they will develop other fast reroute mechanisms. But there are no other LFA mechanisms available. There's only LFA. There's different types of LFA though (link protection, and node protection) but they are not spelled out in FRR. Now the idea is, if there's going to be LFA and TI-LFA as different options then yeah it would make sense. However it seems that they haven't coded out both for both protocols....which is odd.

So I figured, hey lets just remove an entire node because it seems like it's not really useful to call it fast-reroute as LFA is the only fast reroute capability to my understanding for IS-IS and OSPF. Again though, I wish LFA and TI-LFA were both coded in...

That all being said, if you want me to add the node of fast-reroute in there and then nest LFA behind it, no problem and I will make that change. I'm not overly married to how it is as it is now, or with fast-reroute as a node. I just figured to remove a node where it seems like it wasn't useful. But for consistency I can definitely add the fast-reroute.

**Comment by Cheeze-It** [comment]
I added the node fast-reroute. Updated the other places where the path is called. 

I *think* we should be good :)

File: interface-definitions/include/isis/lfa-local.xml.i [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+              </properties>
+            </leafNode>
+          </children>
+        </node>
+        <node name="high">
+          <properties>
+            <help>Compute for critical, and high priority prefixes</help>
+          </properties>
+          <children>
+            <leafNode name="level-1">
```
Comment (c-po) [inline interface-definitions/include/isis/lfa-local.xml.i]: This seems to be a candidate for an XML.I include file

**Review by c-po — APPROVED** [review]

File: interface-definitions/include/isis/lfa-local.xml.i (hunk 1 above)
Comment (Cheeze-It) [reply interface-definitions/include/isis/lfa-local.xml.i]: Made changes. Please let me know if that's what you wanted my good sir :)

**Comment by sever-sever** [comment]
In your example are configured `level-1` and `level-2`

but FRR generates only level-1
```
fast-reroute remote-lfa prefix-list test-lfa-prefix-list level-1
```
Is it ok?

**Comment by Cheeze-It** [comment]
> In your example are configured level-1 and level-2
> 
> but FRR generates only level-1
> 
> fast-reroute remote-lfa prefix-list test-lfa-prefix-list level-1
> 
> is it ok?

Good catch, I....didn't see that, one moment please. I'll fix it. 

Thank you my friend.

**Review by dmbaturin — APPROVED** [review]

**Comment by Cheeze-It** [comment]
@sever-sever, ok my friend I think I got it. I added a protocol handler that will not allow both levels to be configured. So here's how we have it now behavior wise.

"normal" configuration which will match both levels will be configured like this:

```
vyos@vyos:~$ show configuration commands | match isis
set protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1
set protocols isis interface lo
set protocols isis net '49.0001.1921.6800.1001.00'
vyos@vyos:~$ vtysh -c "show run"
Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
 fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-1
 fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-2
exit
!
ip prefix-list lfa-prefix-list-test-1 seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

Level 1 will be added like this:

```
vyos@vyos:~$ configure
WARNING: There was a config error on boot: saving the configuration now could overwrite data.
You may want to check and reload the boot config
[edit]
vyos@vyos# set protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1 level-1
[edit]
vyos@vyos# compare
[protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1]
+ level-1

[edit]
vyos@vyos# commit
_add_log_entry: CRITICAL:[Errno 13] Permission denied: '/opt/vyatta/etc/config/archive/commits'
[edit]
vyos@vyos# exit
Warning: configuration changes have not been saved.
exit

vyos@vyos:~$ vtysh -c "show run"
Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
 fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-1
exit
!
ip prefix-list lfa-prefix-list-test-1 seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

Level 2 will be added like this:

```
vyos@vyos:~$ configure
WARNING: There was a config error on boot: saving the configuration now could overwrite data.
You may want to check and reload the boot config
[edit]
vyos@vyos# delete protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1 level-1
[edit]
vyos@vyos# set protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1 level-2
[edit]
vyos@vyos# compare
[protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1]
- level-1
+ level-2

[edit]
vyos@vyos# commit
_add_log_entry: CRITICAL:[Errno 13] Permission denied: '/opt/vyatta/etc/config/archive/commits'
[edit]
vyos@vyos# exit
Warning: configuration changes have not been saved.
exit
vyos@vyos:~$ vtysh -c "show run"
Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
 fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-2
exit
!
ip prefix-list lfa-prefix-list-test-1 seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

When we try to add both we get the error:

```
vyos@vyos:~$ configure
WARNING: There was a config error on boot: saving the configuration now could overwrite data.
You may want to check and reload the boot config
[edit]
vyos@vyos# set protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1 level-1
[edit]
vyos@vyos# compare
[protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1]
+ level-1

[edit]
vyos@vyos# commit

LFA remote prefix-list lfa-prefix-list-test-1 cannot have both levels
configured at the same time.

[[protocols isis]] failed
Commit failed
_add_log_entry: CRITICAL:[Errno 13] Permission denied: '/opt/vyatta/etc/config/archive/commits'
[edit]
vyos@vyos# exit discard
exit
```

Let me know if that is reasonable.

**Comment by sever-sever** [comment]
There is something wrong with the logic
```
vyos@r4# set protocols isis fast-reroute lfa remote prefix-list test level-1
[edit]
vyos@r4# set protocols isis fast-reroute lfa remote prefix-list test level-2
[edit]
vyos@r4# set protocols isis net "49.0001.1921.6800.1002.00"
[edit]
vyos@r4# commit
[ protocols isis ]
LFA remote prefix-list test cannot have both levels configured at the
same time.

```
Why cannot have both? If it is possible? It is not clear.
```
!
router isis VyOS
 net 49.0001.1921.6800.1002.00
 fast-reroute remote-lfa prefix-list test level-1
 fast-reroute remote-lfa prefix-list test level-2
exit
!

```

**Comment by Cheeze-It** [comment]
In FRR the behavior is like this. If you just add the prefix list it actually adds the configuration for both levels.

Here's my baseline config:

```
vyos(config-router)# do show running-config
Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface eth0
exit
!
interface eth1
exit
!
interface lo
exit
!
router isis test
exit
!
ip prefix-list lfa-prefix-list-test-1 seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```


When I add the config for the prefix list only I get:

```
vyos(config-router)# fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1
vyos(config-router)#
vyos(config-router)#
vyos(config-router)#
vyos(config-router)# do show running-config
Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface eth0
exit
!
interface eth1
exit
!
interface lo
exit
!
router isis test
 fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-1
 fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-2
exit
!
ip prefix-list lfa-prefix-list-test-1 seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

The equivalent to that for the additions to the CLI in VyOS are:

```
FRR: fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1

VyOS: set protocols isis fast-reroute lfa remote prefix-list lfa-prefix-list-test-1
```

FRR also lets you specify both levels as well, but that is redundant if all you need to do is specify the prefix list:

```
vyos(config-router)# fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-1
vyos(config-router)# fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-2
vyos(config-router)#
vyos(config-router)# do show running-config
Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface eth0
exit
!
interface eth1
exit
!
interface lo
exit
!
router isis test
 fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-1
 fast-reroute remote-lfa prefix-list lfa-prefix-list-test-1 level-2
exit
!
ip prefix-list lfa-prefix-list-test-1 seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

I was thinking to myself, why not just make it easier in VyOS, I'll follow the default behavior of FRR. The default behavior is, if you specify the prefix list then it applies that prefix list to both levels. What I specifically did here was I removed the ability in VyOS (via a config check) to allow for one to configure both levels along with a configured prefix list. One can still enable a singular level if one is to match to a singular level. I figure it would be confusing to allow for the configuration to allow one to enable the prefix list *and* allow for both levels to be configured at the same time. However if you'd like, what we can do is require for a prefix list to have one or more levels configured.

**Comment by Cheeze-It** [comment]
Alrighty, I think I got the behavior working now.

We're starting with this config:

```
vyos@vyos:~$ show configuration commands | match isis
set protocols isis interface lo
set protocols isis net '49.0001.1921.6800.1001.00'

Every 1.0s: /usr/bin/vtysh -c "show run"                                                                                                                                                                                                                                                     vyos: Fri Oct  6 06:10:10 2023

Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
exit
!
ip prefix-list testing seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

Adding the main configuration without a level and we get:

```
vyos@vyos:~$ configure
[edit]
vyos@vyos# set protocols isis fast-reroute lfa remote prefix-list testing
[edit]
vyos@vyos# compare
[protocols isis]
+ fast-reroute {
+     lfa {
+         remote {
+             prefix-list testing {
+             }
+         }
+     }
+ }

[edit]
vyos@vyos# commit
_add_log_entry: CRITICAL:[Errno 13] Permission denied: '/opt/vyatta/etc/config/archive/commits'
[edit]
vyos@vyos#

Every 1.0s: /usr/bin/vtysh -c "show run"                                                                                                                                                                                                                                                     vyos: Fri Oct  6 06:11:59 2023

Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
 fast-reroute remote-lfa prefix-list testing level-1
 fast-reroute remote-lfa prefix-list testing level-2
exit
!
ip prefix-list testing seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

Adding level 1 only:

```
[edit]
vyos@vyos# set protocols isis fast-reroute lfa remote prefix-list testing level-1
[edit]
vyos@vyos# compare
[protocols isis fast-reroute lfa remote prefix-list testing]
+ level-1

[edit]
vyos@vyos# commit
_add_log_entry: CRITICAL:[Errno 13] Permission denied: '/opt/vyatta/etc/config/archive/commits'
[edit]
vyos@vyos#

Every 1.0s: /usr/bin/vtysh -c "show run"                                                                                                                                                                                                                                                     vyos: Fri Oct  6 06:13:26 2023

Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
 fast-reroute remote-lfa prefix-list testing level-1
exit
!
ip prefix-list testing seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

Adding level 2 only:

```
[edit]
vyos@vyos# delete protocols isis fast-reroute lfa remote prefix-list testing level-1
[edit]
vyos@vyos# set protocols isis fast-reroute lfa remote prefix-list testing level-2
[edit]
vyos@vyos# compare
[protocols isis fast-reroute lfa remote prefix-list testing]
- level-1
+ level-2

[edit]
vyos@vyos# commit
_add_log_entry: CRITICAL:[Errno 13] Permission denied: '/opt/vyatta/etc/config/archive/commits'
[edit]

Every 1.0s: /usr/bin/vtysh -c "show run"                                                                                                                                                                                                                                                     vyos: Fri Oct  6 06:14:12 2023

Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
 fast-reroute remote-lfa prefix-list testing level-2
exit
!
ip prefix-list testing seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

Adding both levels now:

```
vyos@vyos# set protocols isis fast-reroute lfa remote prefix-list testing level-1
[edit]
vyos@vyos# compare
[protocols isis fast-reroute lfa remote prefix-list testing]
+ level-1

[edit]
vyos@vyos# commit
_add_log_entry: CRITICAL:[Errno 13] Permission denied: '/opt/vyatta/etc/config/archive/commits'
[edit]

Every 1.0s: /usr/bin/vtysh -c "show run"                                                                                                                                                                                                                                                     vyos: Fri Oct  6 06:14:51 2023

Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
 fast-reroute remote-lfa prefix-list testing level-1
 fast-reroute remote-lfa prefix-list testing level-2
exit
!
ip prefix-list testing seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

The final config looks like this in VyOS:

```
vyos@vyos:~$ show configuration commands | match isis
set protocols isis fast-reroute lfa remote prefix-list testing level-1
set protocols isis fast-reroute lfa remote prefix-list testing level-2
set protocols isis interface lo
set protocols isis net '49.0001.1921.6800.1001.00'
vyos@vyos:~$

vyos@vyos:~$ vtysh -c "show run"
Building configuration...

Current configuration:
!
frr version 9.0
frr defaults traditional
hostname vyos
log syslog
log facility local7
service integrated-vtysh-config
!
ip route 0.0.0.0/0 10.0.0.65
!
interface lo
 ip router isis VyOS
 ipv6 router isis VyOS
exit
!
router isis VyOS
 net 49.0001.1921.6800.1001.00
 fast-reroute remote-lfa prefix-list testing level-1
 fast-reroute remote-lfa prefix-list testing level-2
exit
!
ip prefix-list testing seq 1 permit 192.168.255.255/32
!
rpki
exit
!
end
```

So I think now @sever-sever it is as you were wanting it to be, so that's good :)

One thing I do want to give a heads up to for everyone here though. The smoketest takes a long time (mine measured running 8 tests in 53.985s) and I can try to make the time lower if you guys wish. The reason I did the test as I did it in said smoketest is because with LFA there's a lot of mutually exclusive settings that tend to only like being tested one at a time. So I just wanted to make sure all of them worked properly. But if need be I can try to make it faster.

**Review by sever-sever — APPROVED** [review]

**Comment by c-po** [comment]
@MergifyIo backport sagitta

**Pull request merged**